lib/delete.ts

```diff
@@ -0,0 +1,286 @@
+import type { WriteRequest, DeleteItemInput } from 'aws-sdk/clients/dynamodb';
+import type { Facet } from './facet';
+import { wait } from './wait';
+import { Converter } from '@faceteer/converter';
+import expressionBuilder from '@faceteer/expression-builder';
+import { PK, SK } from './keys';
+
+export interface DeleteOptions<T> {
+	condition?: expressionBuilder.ConditionExpression<T>;
+}
+
+/**
+ * A failed delete request
+ */
+export interface DeleteFailure<T> {
+	/**
+	 * The record that failed to be deleted
+	 */
+	record: T;
+	/**
+	 * The error from trying to delete the item
+	 */
+	error: unknown;
+}
+
+/**
+ * A response from a delete request to a Facet
+ */
+export interface DeleteResponse<T> {
+	/**
+	 * If the request had any failures
+	 */
+	hasFailures: boolean;
+	/**
+	 * The records that were successfully deleted
+	 */
+	deleted: T[];
+	/**
+	 * The records that failed to be deleted
+	 */
+	failed: DeleteFailure<T>[];
+}
+
+export async function deleteSingleItem<
+	T,
+	PK extends keyof T,
+	SK extends keyof T,
+	U = Pick<T, PK | SK> & Partial<T>,
+>(
+	facet: Facet<T, PK, SK>,
+	record: U,
+	options: DeleteOptions<T> = {},
+): Promise<DeleteResponse<U>> {
+	try {
+		const deleteInput: DeleteItemInput = {
+			TableName: facet.connection.tableName,
+			Key: {
+				[PK]: {
+					S: facet.pk(record),
+				},
+				[SK]: {
+					S: facet.sk(record),
+				},
+			},
+		};
+
+		if (options.condition) {
+			const expression = expressionBuilder.condition(options.condition);
+			deleteInput.ConditionExpression = expression.expression;
+			deleteInput.ExpressionAttributeNames = expression.names;
+			deleteInput.ExpressionAttributeValues = expression.values;
+		}
+
+		const response = await facet.connection.dynamoDb
+			.deleteItem(deleteInput)
+			.promise();
+
+		if (response.$response.error) {
+			return {
+				failed: [
+					{
+						error: response.$response.error,
+						record,
+					},
+				],
+				hasFailures: true,
+				deleted: [],
+			};
+		}
+
+		return {
+			failed: [],
+			hasFailures: false,
+			deleted: [record],
+		};
+	} catch (error) {
+		return {
+			failed: [
+				{
+					error: error,
+					record,
+				},
+			],
+			hasFailures: true,
+			deleted: [],
+		};
+	}
+}
+
+/**
+ * Delete records from the from Dynamo DB
+ * @param records
+ */
+export async function deleteItems<
+	T,
+	PK extends keyof T,
+	SK extends keyof T,
+	U = Pick<T, PK | SK> & Partial<T>,
+>(facet: Facet<T, PK, SK>, records: U[]): Promise<DeleteResponse<U>> {
+	const recordsToBatch: U[] = [...records];
+	const deleteResponse: DeleteResponse<U> = {
+		failed: [],
+		hasFailures: false,
+		deleted: [],
+	};
+	/**
+	 * Dynamo DB only allows 25 items in a write batch
+	 * request so we will break this down into batches
+	 */
+	const batches: U[][] = [];
+
+	while (recordsToBatch.length >= 1) {
```

Comment:
Shouldn't this be like below?
```suggestion
	while (recordsToBatch.length >= 25) {
```

Reply:
I understand now.